examples/ExactlyOnce/Program.cs

```diff
@@ -15,788 +15,988 @@
+using System.IO;
+
+
 namespace Confluent.Kafka.Examples.ExactlyOnce
 {
     public class Program
+    {
+        private static readonly ILogger logger = Utils.LoggerFactory.CreateLogger<Program>();
+
+        public static async Task Main(string[] args)
+        {
+            var cts = new CancellationTokenSource();
+            var completedCancellationTokenSource = new CancellationTokenSource();
+            Console.CancelKeyPress += (_, e) =>
+            {
+                e.Cancel = true; // prevent the process from terminating.
+                cts.Cancel();
+            };
+            AppDomain.CurrentDomain.ProcessExit += (_, e) =>
+            {
+                cts.Cancel();
+                completedCancellationTokenSource.Token.WaitHandle.WaitOne();
+            };
+
+            if (args.Length == 0 ||
+                (args[0] == "print_words" && args.Length != 1) ||
+                args[0] != "print_words" && args.Length != 2)
+            {
+                Console.WriteLine("Usage:");
+                Console.WriteLine("  create_words <transactional-id>");
+                Console.WriteLine("  reverse_words <transactional-id>");
+                Console.WriteLine("  print_words");
+                completedCancellationTokenSource.Cancel();
+                return;
+            }
+
+            var command = args[0];
+            string transactionalId = null;
+            if (args.Length > 1) transactionalId = args[1];
+
+            IConfiguration configuration = new ConfigurationBuilder()
+                .SetBasePath(Directory.GetCurrentDirectory())
+                .AddJsonFile("./appsettings.json")
+                .Build();
+
+            var adminClientConfig = configuration.GetSection("AdminClient").Get<AdminClientConfig>();
+            using var adminClient = new AdminClientBuilder(adminClientConfig).Build();
+
+            switch (command)
+            {
+                case "create_words":
+                    await new WordCreator().Execute(configuration, transactionalId, adminClient, cts);
+                    break;
+                case "reverse_words":
+                    await new WordReverser().Execute(configuration, transactionalId, adminClient, cts);
+                    break;
+                case "print_words":
+                    await new WordPrinter().Execute(configuration, adminClient, cts);
+                    break;
+                default:
+                    logger.LogError("Unknown command");
+                    break;
+            }
+
+            completedCancellationTokenSource.Cancel();
+        }
+    }
+
+    /// <summary>
+    ///     Error that is thrown if the transactional producer receives a
+    ///     fatal error.
+    /// </summary>
+    public class TransactionalProcessorFatalError : Exception
+    {
+    }
+
+    /// <summary>
+    ///     Represents a single message to be sent
+    ///     to an output topic.
+    /// </summary>
+    public class ProduceMessage<K, V>
```

Comment:
I think it's needed, imagine you want to do transactions involving multiple input and multiple output topics. It's common enough.
About the delivery handler, if you have a producer only and want to do transactions involving messages sent from multiple http calls, let's says many users belonging to the same group that have to be sent all together or none, you may want to give them a response when complete. You could send everyone the same response when EndTransaction is called, but can be useful to have the offset, partition or error code for logging. Also it's optional so can be used only if needed

Reply:
But what happens if the transaction is aborted? I'm thinking that it could be changed to use ProduceAsync, get the delivery reports, await them at the end of transaction and pass them to EndTransaction to avoid having the caller wait for EndTransaction to actually deliver those reports.

Reply:
> I think it's needed, imagine you want to do transactions involving multiple input and multiple output topics. It's common enough.

yes. i thought i'd thought of a better way to do it (using the return mechanism), but i can't think of what i might have been thinking now. A significant issue with what you propose is you don't want to be specifying `Topic = config.OutputTopic.Name,` - kind of defeats the purpose of having OutputTopic.

perhaps we don't want to return the result after all, perhaps we want a `Produce` method for this. but rather than exposing the producer instance to the user and have the user call `Produce` on that, we keep that encapsulated, and put a Produce method on the `Processor` class.

I don't quite see right now how the delivery reports work. Maybe they aren't exposed by the new `Produce` method, but maybe they are made available at transaction end. Since delivery reports contain the message instance, and all the meta info, everything is there to satisfy the use case you point out i think.

Feel free to iterate on the above idea and see if it works - just an idea. I'm convinced that having `ProduceMessage` is going to lead to a worse structure than optimal.